rolling-update-kubernetes/src/main/scala/akka/rollingupdate/kubernetes/PodDeletionCostAnnotator.scala

```diff
@@ -0,0 +1,211 @@
+    val factory = KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm)
+    val keyStore = KeyStore.getInstance("PKCS12")
+    keyStore.load(null)
+    factory.init(keyStore, Array.empty)
+    val km: Array[KeyManager] = factory.getKeyManagers
+    val tm: Array[TrustManager] =
+      PemManagersProvider.buildTrustManagers(certificates)
+    val random: SecureRandom = new SecureRandom
+    val sslContext = SSLContext.getInstance("TLSv1.2")
+    sslContext.init(km, tm, random)
+    sslContext
+  }
+  private val clientSslContext: Option[HttpsConnectionContext] =
+    if (settings.secure) Some(ConnectionContext.httpsClient(sslContext)) else None
+
+  implicit val dispatcher: ExecutionContextExecutor = context.system.dispatcher
+  def receive = idle(0, SortedSet.empty(Member.ageOrdering))
+
+  private def idle(deletionCost: Int, membersByAgeDesc: SortedSet[Member]): Receive = {
+    case cs @ ClusterEvent.CurrentClusterState(members, _, _, _, _) =>
+      log.debug("Received CurrentClusterState {}", cs)
+      updateIfNewCost(deletionCost, membersByAgeDesc ++ members) // ordering used is from the first operand (so, by age)
+
+    case ClusterEvent.MemberUp(m) =>
+      log.debug("Received MemberUp {}", m)
+      updateIfNewCost(deletionCost, membersByAgeDesc + m)
+
+    case ClusterEvent.MemberRemoved(m, _) =>
+      log.debug("Received MemberRemoved {}", m)
+      updateIfNewCost(deletionCost, membersByAgeDesc - m)
+
+    case PodAnnotated =>
+      log.debug("Annotation updated successfully to {}", deletionCost)
+      // cancelling an eventual retry in case the annotation succeeded in the meantime
+      timers.cancel(RetryTimerId)
+
+    case ScheduleRetry(ex) =>
+      log.error(s"Failed to update annotation: [$ex]. Retrying with fixed delay of {}.", costSettings.retryDelay)
+      timers.startSingleTimer(RetryTimerId, RetryAnnotate, costSettings.retryDelay)
+      context.become(underRetryBackoff(membersByAgeDesc))
+
+    case GiveUp(er: String) =>
+      log.error(
+        "There was a client error when trying to set pod-deletion-cost annotation. " +
+        "Not retrying, check configuration. Error: {}",
+        er)
+
+    case es => log.debug("Ignoring message {}", es)
+  }
+
+  private def underRetryBackoff(membersByAgeDesc: SortedSet[Member]): Receive = {
+    case ClusterEvent.MemberUp(m) =>
+      log.debug("Received while on retry backoff MemberUp {}", m)
+      context.become(underRetryBackoff(membersByAgeDesc + m))
+
+    case ClusterEvent.MemberRemoved(m, _) =>
+      log.debug("Received while on retry backoff MemberRemoved {}", m)
+      context.become(underRetryBackoff(membersByAgeDesc - m))
+
+    case RetryAnnotate =>
+      updateIfNewCost(Int.MinValue, membersByAgeDesc)
+
+    case es => log.debug("Under retry backoff, ignoring message {}", es)
+  }
+
+  private def updateIfNewCost(existingCost: Int, membersByAgeDesc: immutable.SortedSet[Member]): Unit = {
```

Comment:
Making sure I got your idea right.. 

Currently, it logs an error every time it fails to update the annotation and schedules a retry.. we could downgrade the severity to warning after say 3 atttempts and also not log it on every attempt but maybe every 10th retry attempt for instance? This is would avoid clogging the logs with the error too much..

Reply:
We talked but to get the clarification in print: the idea was to log at debug or info first N tries as that is just resilience/retrying, but after N tries it might be a sign of misconfiguration, so start log at warn. This is probably not critical enough to ever be an error though the deployed app will work even if this would not work.